File: src/naming.py
"""Utility helpers to standardise output filenames.

These functions centralise how dataset tags and figure paths are
constructed so that both Python and MATLAB scripts produce
consistent results.
"""
from __future__ import annotations

from pathlib import Path


def build_tag(imu_file: str, gnss_file: str, method: str) -> str:
    """Return dataset tag ``IMU_Xnnn_GNSS_Xnnn_METHOD``."""
    imu_base = Path(imu_file).stem
    gnss_base = Path(gnss_file).stem
    return f"{imu_base}_{gnss_base}_{method}"


def prefix_filename(tag: str | None, filename: str) -> str:
    """Prepend ``tag`` to ``filename`` if provided."""
    return f"{tag}_{filename}" if tag else filename


def plot_path(
    results_dir: str | Path,
    tag: str,
    task: int,
    subtask: str,
    output_type: str,
    ext: str = ".pdf",
) -> Path:
    """Compose a plot filename following the repository convention."""
    results_dir = Path(results_dir)
    fname = f"{tag}_task{task}_{subtask}_{output_type}{ext}"
    return results_dir / fname

File: MATLAB/src/naming.m
function path = naming(varargin)
%NAMING Utility helpers for standardised filenames.
%   TAG = NAMING('build_tag', IMU_FILE, GNSS_FILE, METHOD) returns a dataset
%   tag such as 'IMU_X001_GNSS_X001_TRIAD'.
%   FILE = NAMING('plot_path', DIR, TAG, TASK, SUBTASK, TYPE, EXT) builds a
%   full path for plots or results.
%
%   Example:
%       tag = naming('build_tag', 'IMU_X001.dat', 'GNSS_X001.csv', 'TRIAD');
%       f = naming('plot_path', 'results', tag, 7, '3', 'residuals_position_velocity');

op = varargin{1};
if strcmp(op, 'build_tag')
    imu_file = varargin{2};
    gnss_file = varargin{3};
    method = varargin{4};
    [~, ibase, ~] = fileparts(imu_file);
    [~, gbase, ~] = fileparts(gnss_file);
    path = sprintf('%s_%s_%s', ibase, gbase, method);
elseif strcmp(op, 'plot_path')
    dir = varargin{2};
    tag = varargin{3};
    task = varargin{4};
    subtask = varargin{5};
    type = varargin{6};
    if nargin < 7
        ext = '.pdf';
    else
        ext = varargin{7};
    end
    fname = sprintf('%s_task%d_%s_%s%s', tag, task, subtask, type, ext);
    path = fullfile(dir, fname);
else
    error('Unknown operation');
end
end
